mod/bookmarks/start.php

```diff
@@ -129,7 +129,7 @@ function bookmarks_page_handler($page) {
  */
 function bookmark_set_url($hook, $type, $url, $params) {
```

Comment:
why not move this function into the ElggBookmark::getURL()

Reply:
I am against adding logic to methods that trigger hooks. We should make everything that calls a hook final, IMO, includign `getURL`, `getIconURL`, `canEdit` etc

Reply:
disagree, but i get your issue with the hooks, maybe we should have an extra function like `prepareGetUrl` that only does the trigger... that way class function can just call `return prepareGetUrl("myurl")`..

`prepareGetUrl` only deals with the trigger... and can be final... it makes no sense to have all these hook functions... makes it also harder to test... class functions are easier

Reply:
Testability vs extendability. I have already had to deal with some of your plugins setting the URL in the object class, which makes it impossible to change the URL

Reply:
i understand the problem, but another approach could solve both challenges

Reply:
Hooks ARE the solution to both challenges

Reply:
i do not want to drop hooks, but i also want to use class functions for getURL... my suggestion is to have both in a simple way